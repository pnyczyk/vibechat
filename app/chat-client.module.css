.main {
  min-height: 100vh;
  width: 100%;
  display: flex;
  justify-content: center;
  padding: clamp(2rem, 6vw, 4rem);
  background-color: #ffffff;
}

.layout {
  display: grid;
  grid-template-columns: minmax(0, 1fr) 260px;
  gap: clamp(2rem, 4vw, 3rem);
  width: min(1100px, 100%);
  min-height: clamp(28rem, 70vh, 42rem);
}

.canvas {
  display: flex;
  flex-direction: column;
  gap: clamp(1.5rem, 2vw, 2.5rem);
  padding: clamp(2.5rem, 5vw, 3.5rem);
  border-radius: 24px;
  border: 1px solid rgba(15, 23, 42, 0.08);
  background-color: #ffffff;
  box-shadow: 0 32px 64px rgba(15, 23, 42, 0.08);
}

.header {
  display: flex;
  flex-direction: column;
  gap: 0.75rem;
  max-width: 32rem;
}

.title {
  font-weight: 600;
}

.surface {
  flex: 1;
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 18px;
  border: 1px dashed rgba(148, 163, 184, 0.8);
  background: rgba(241, 245, 249, 0.45);
  color: rgba(71, 85, 105, 0.9);
  text-align: center;
  padding: 1.5rem;
}

.status {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
}

.controlRail {
  display: flex;
  justify-content: flex-end;
  align-items: stretch;
}

.controlRailInner {
  position: sticky;
  top: clamp(2rem, 6vw, 4rem);
  display: flex;
  flex-direction: column;
  gap: 1.5rem;
  width: min(100%, 16rem);
  padding: 1.75rem;
  border-radius: 24px;
  border: 1px solid rgba(226, 232, 240, 1);
  background: rgba(241, 245, 249, 0.8);
  box-shadow: 0 24px 48px rgba(15, 23, 42, 0.05);
}

.controlHeading {
  font-size: 0.875rem;
  letter-spacing: 0.08em;
  text-transform: uppercase;
  font-weight: 600;
  color: rgba(71, 85, 105, 0.9);
}

.controlStack {
  display: flex;
  flex-direction: column;
  gap: 1rem;
}

@media (max-width: 1024px) {
  .layout {
    grid-template-columns: 1fr;
  }

  .controlRail {
    justify-content: center;
  }

  .controlRailInner {
    position: static;
    flex-direction: row;
    align-items: center;
    justify-content: center;
    gap: 1.25rem;
    width: min(100%, 28rem);
    border-radius: 999px;
    padding: 1rem 1.5rem;
  }

  .controlStack {
    flex-direction: row;
    gap: 1rem;
  }
}

@media (max-width: 640px) {
  .main {
    padding: clamp(1.5rem, 6vw, 2rem);
  }

  .canvas {
    padding: clamp(1.75rem, 6vw, 2.25rem);
  }

  .controlRailInner {
    flex-direction: column;
    border-radius: 24px;
    width: 100%;
  }

  .controlStack {
    flex-direction: column;
    width: 100%;
  }
}
