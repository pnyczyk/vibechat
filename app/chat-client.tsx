"use client";

import { useCallback, useEffect, useMemo, useState } from "react";
import Typography from "@mui/material/Typography";
import { RealtimeAgent, RealtimeSession } from "@openai/agents/realtime";
import styles from "./chat-client.module.css";

import {
  SessionControls,
  SessionFeedback,
  ConnectionStatus,
} from "./components/SessionControls";

export function ChatClient() {
  const [status, setStatus] = useState<ConnectionStatus>("idle");
  const [error, setError] = useState<string | null>(null);
  const [feedback, setFeedback] = useState<SessionFeedback | null>(null);
  const [session, setSession] = useState<RealtimeSession | null>(null);

  useEffect(() => {
    return () => {
      session?.close();
    };
  }, [session]);

  const agent = useMemo(() => {
    return new RealtimeAgent({
      name: "Assistant",
      instructions: "Mów po polsku i odpowiadaj głosem. Bądź serdeczny.",
    });
  }, []);

  const handleConnect = useCallback(async () => {
    if (status === "connecting" || status === "connected") {
      return;
    }

    if (session) {
      session.close();
      setSession(null);
    }

    setError(null);
    setFeedback(null);
    setStatus("connecting");

    try {
      const response = await fetch("/api/realtime-token");
      if (!response.ok) {
        const body = await response.json().catch(() => ({}));
        const message =
          typeof body.error === "string"
            ? body.error
            : `Token endpoint returned ${response.status}`;
        throw new Error(message);
      }

      const secret = await response.json();
      const apiKey =
        typeof secret?.value === "string"
          ? secret.value
          : secret?.client_secret?.value;
      if (!apiKey) {
        throw new Error("Realtime token response missing value");
      }

      const newSession = new RealtimeSession(agent, {
        model: "gpt-realtime",
      });

      await newSession.connect({ apiKey });
      setSession(newSession);
      setStatus("connected");
      setFeedback({ message: "Connected to session", severity: "success" });
    } catch (err) {
      console.error("Failed to connect realtime session", err);
      setStatus("error");
      const message = err instanceof Error ? err.message : "Unexpected error";
      setError(message);
      setFeedback({ message, severity: "error" });
    }
  }, [agent, session, status]);

  const handleDisconnect = useCallback(() => {
    session?.close();
    setSession(null);
    setStatus("idle");
    setError(null);
    setFeedback({ message: "Disconnected from session", severity: "success" });
  }, [session]);

  const handleFeedbackClose = useCallback(() => {
    setFeedback(null);
  }, []);

  return (
    <section className={styles.layout} aria-labelledby="chat-title">
      <div className={styles.canvas}>
        <header className={styles.header}>
          <Typography
            id="chat-title"
            variant="h3"
            component="h1"
            className={styles.title}
          >
            VibeChat
          </Typography>
          <Typography variant="body1" color="text.secondary">
            Connect to start a realtime voice session or explore the workspace
            while we prepare new modules.
          </Typography>
        </header>

        <div className={styles.surface} role="presentation">
          <Typography variant="body2" color="text.secondary">
            Voice interaction canvas reserved for upcoming live session
            visualization.
          </Typography>
        </div>

        <footer className={styles.status} aria-live="polite">
          <Typography variant="body2">Status: {status}</Typography>
          {error ? (
            <Typography variant="body2" color="error">
              Error: {error}
            </Typography>
          ) : (
            <Typography variant="body2" color="text.secondary">
              Allow microphone access when prompted to keep the session ready.
            </Typography>
          )}
        </footer>
      </div>

      <aside className={styles.controlRail} aria-label="Session controls">
        <div className={styles.controlRailInner}>
          <SessionControls
            status={status}
            onConnect={handleConnect}
            onDisconnect={handleDisconnect}
            feedback={feedback}
            onFeedbackClose={handleFeedbackClose}
          />
        </div>
      </aside>
    </section>
  );
}
