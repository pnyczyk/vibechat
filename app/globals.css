@import "katex/dist/katex.min.css";

:root {
  color-scheme: var(--vc-color-scheme, light);
}

* {
  box-sizing: border-box;
}

html,
body {
  margin: 0;
  width: 100%;
  height: 100%;
}

body.vc-body {
  display: flex;
  flex-direction: column;
  min-height: 100vh;
  min-width: 100vw;
  width: 100%;
  background-color: var(--vc-background, #ffffff);
  overflow: hidden;
}

a {
  color: inherit;
  text-decoration: none;
}
