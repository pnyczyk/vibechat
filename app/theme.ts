import { createTheme } from '@mui/material/styles';

const theme = createTheme({
  palette: {
    mode: 'light',
    background: {
      default: '#ffffff',
      paper: '#ffffff',
    },
  },
  typography: {
    fontFamily: 'Roboto, Helvetica, Arial, sans-serif',
  },
  components: {
    MuiCssBaseline: {
      styleOverrides: {
        body: {
          backgroundColor: '#ffffff',
        },
      },
    },
  },
});

export default theme;
