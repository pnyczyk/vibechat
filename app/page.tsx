import { ChatClient } from './chat-client';
import styles from './chat-client.module.css';

export default function HomePage() {
  return (
    <main className={styles.main}>
      <ChatClient />
    </main>
  );
}
