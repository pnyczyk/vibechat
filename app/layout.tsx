import type { Metadata } from 'next';
import React from 'react';
import './globals.css';

export const metadata: Metadata = {
  title: 'VibeChat',
  description: 'Realtime assistant demo powered by OpenAI agents.',
};

export default function RootLayout({ children }: { children: React.ReactNode }) {
  return (
    <html lang="en">
      <body>{children}</body>
    </html>
  );
}
