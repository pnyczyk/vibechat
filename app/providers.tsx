"use client";

import { CssBaseline, ThemeProvider } from "@mui/material";
import type { PropsWithChildren } from "react";
import { createContext, useCallback, useContext, useEffect, useMemo, useRef, useState } from "react";
import { createAppTheme } from "./theme";
import { ThemeStore, type ThemeMode } from "./lib/theme-store";

type ThemeController = {
  mode: ThemeMode;
  toggle: () => void;
  setMode: (mode: ThemeMode) => void;
};

const ThemeControllerContext = createContext<ThemeController | null>(null);

export const useThemeController = () => {
  const context = useContext(ThemeControllerContext);
  if (!context) {
    throw new Error("useThemeController must be used within Providers");
  }
  return context;
};

export default function Providers({ children }: PropsWithChildren) {
  const themeStoreRef = useRef<ThemeStore | null>(null);

  if (!themeStoreRef.current) {
    themeStoreRef.current = new ThemeStore();
  }

  const themeStore = themeStoreRef.current!;
  const [mode, setMode] = useState<ThemeMode>(() => themeStore.getMode());

  useEffect(() => {
    const unsubscribe = themeStore.subscribe((nextMode) => {
      setMode(nextMode);
    });

    themeStore.hydrate();
    return unsubscribe;
  }, [themeStore]);

  const handleToggle = useCallback(() => {
    themeStore.toggle();
  }, [themeStore]);

  const handleSetMode = useCallback((nextMode: ThemeMode) => {
    themeStore.setMode(nextMode);
  }, [themeStore]);

  const theme = useMemo(() => createAppTheme(mode), [mode]);

  const controller = useMemo(
    () => ({
      mode,
      toggle: handleToggle,
      setMode: handleSetMode,
    }),
    [handleSetMode, handleToggle, mode],
  );

  useEffect(() => {
    if (typeof document === "undefined") {
      return;
    }

    document.documentElement.style.setProperty("--vc-color-scheme", mode);
    document.documentElement.style.colorScheme = mode;
    const background = theme.palette.background.default;
    document.body.style.setProperty("--vc-background", background);
    document.body.style.backgroundColor = background;
    document.documentElement.style.backgroundColor = background;
  }, [mode, theme]);

  return (
    <ThemeControllerContext.Provider value={controller}>
      <ThemeProvider theme={theme}>
        <CssBaseline />
        {children}
      </ThemeProvider>
    </ThemeControllerContext.Provider>
  );
}
