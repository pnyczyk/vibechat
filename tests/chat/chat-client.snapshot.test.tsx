import { render, screen } from "@testing-library/react";
import React from "react";
import Providers from "../../app/providers";
import { ChatClient } from "../../app/chat-client";

jest.mock("@openai/agents/realtime", () => {
  class MockRealtimeSession {
    connect = jest.fn();
    close = jest.fn();
    mute = jest.fn();
    history: unknown[] = [];
    on = jest.fn();
    off = jest.fn();
    getLatestAudioLevel = jest.fn().mockReturnValue(null);
  }

  return {
    RealtimeAgent: jest.fn().mockImplementation(() => ({})),
    RealtimeSession: jest
      .fn()
      .mockImplementation(() => new MockRealtimeSession()),
    OpenAIRealtimeWebRTC: jest.fn().mockImplementation(() => ({})),
  };
});

describe("ChatClient layout", () => {
  it("renders control rail and voice activity indicator", () => {
    render(
      <Providers>
        <ChatClient />
      </Providers>,
    );

    expect(screen.getByRole("heading", { name: /vibechat/i })).toBeInTheDocument();
    const indicator = screen.getByTestId("voice-activity-indicator");
    expect(indicator).toHaveAttribute("aria-label", expect.stringMatching(/waiting for audio/i));
    expect(screen.getByTestId("session-controls")).toContainElement(indicator);
  });
});
