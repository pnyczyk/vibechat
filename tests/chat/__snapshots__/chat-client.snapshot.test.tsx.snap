// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ChatClient layout matches the minimalist canvas snapshot 1`] = `
.emotion-0 {
  margin: 0;
  font-family: Roboto,Helvetica,Arial,sans-serif;
  font-weight: 400;
  font-size: 3rem;
  line-height: 1.167;
}

.emotion-1 {
  margin: 0;
  font-family: Roboto,Helvetica,Arial,sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  color: rgba(0, 0, 0, 0.6);
}

.emotion-2 {
  margin: 0;
  font-family: Roboto,Helvetica,Arial,sans-serif;
  font-weight: 400;
  font-size: 0.875rem;
  line-height: 1.43;
  color: rgba(0, 0, 0, 0.6);
}

.emotion-3 {
  margin: 0;
  font-family: Roboto,Helvetica,Arial,sans-serif;
  font-weight: 400;
  font-size: 0.875rem;
  line-height: 1.43;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-5>:not(style):not(style) {
  margin: 0;
}

.emotion-5>:not(style)~:not(style) {
  margin-top: 4px;
}

.emotion-6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.5rem;
  padding: 8px;
  border-radius: 50%;
  color: rgba(0, 0, 0, 0.54);
  -webkit-transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  --IconButton-hoverBg: rgba(0, 0, 0, 0.04);
  color: #1976d2;
  --IconButton-hoverBg: rgba(25, 118, 210, 0.04);
  padding: 12px;
  font-size: 1.75rem;
}

.emotion-6::-moz-focus-inner {
  border-style: none;
}

.emotion-6.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-6 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-6:hover {
  background-color: var(--IconButton-hoverBg);
}

@media (hover: none) {
  .emotion-6:hover {
    background-color: transparent;
  }
}

.emotion-6.Mui-disabled {
  background-color: transparent;
  color: rgba(0, 0, 0, 0.26);
}

.emotion-6.MuiIconButton-loading {
  color: transparent;
}

.emotion-7 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  fill: currentColor;
  font-size: inherit;
}

<section
  aria-labelledby="chat-title"
  class="layout"
>
  <div
    class="canvas"
  >
    <header
      class="header"
    >
      <h1
        class="MuiTypography-root MuiTypography-h3 title emotion-0"
        id="chat-title"
      >
        VibeChat
      </h1>
      <p
        class="MuiTypography-root MuiTypography-body1 emotion-1"
      >
        Connect to start a realtime voice session or explore the workspace while we prepare new modules.
      </p>
    </header>
    <div
      class="surface"
      role="presentation"
    >
      <p
        class="MuiTypography-root MuiTypography-body2 emotion-2"
      >
        Voice interaction canvas reserved for upcoming live session visualization.
      </p>
    </div>
    <footer
      aria-live="polite"
      class="status"
    >
      <p
        class="MuiTypography-root MuiTypography-body2 emotion-3"
      >
        Status: 
        idle
      </p>
      <p
        class="MuiTypography-root MuiTypography-body2 emotion-2"
      >
        Allow microphone access when prompted to keep the session ready.
      </p>
    </footer>
  </div>
  <aside
    aria-label="Session controls"
    class="controlRail"
  >
    <div
      class="controlRailInner"
    >
      <div
        class="MuiStack-root emotion-5"
        data-testid="session-controls"
      >
        <span
          aria-label="Connect to session"
          class=""
          data-mui-internal-clone-element="true"
        >
          <button
            aria-label="Connect to session"
            aria-pressed="false"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeLarge emotion-6"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit emotion-7"
              data-testid="PowerSettingsNewIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M13 3h-2v10h2zm4.83 2.17-1.42 1.42C17.99 7.86 19 9.81 19 12c0 3.87-3.13 7-7 7s-7-3.13-7-7c0-2.19 1.01-4.14 2.58-5.42L6.17 5.17C4.23 6.82 3 9.26 3 12c0 4.97 4.03 9 9 9s9-4.03 9-9c0-2.74-1.23-5.18-3.17-6.83"
              />
            </svg>
          </button>
        </span>
      </div>
    </div>
  </aside>
</section>
`;
